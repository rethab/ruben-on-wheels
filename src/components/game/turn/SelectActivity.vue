<template>
  <v-card>
    <v-card-title>{{ player.name }}'s turn</v-card-title>
    <CardSubtitle :player="player" />

    <v-card-text>
      <p v-html="actionText" />
      <p class="my-5">How do you want to proceed?</p>

      <v-radio-group>
        <v-radio
          :disabled="!selectableActivity(activity)"
          v-for="(activity, index) in cityActivities"
          :key="index"
          :label="activityLabel(activity)"
          :value="activity.name"
          @change="selectedActivity = activity"
        ></v-radio>
      </v-radio-group>
    </v-card-text>

    <v-card-actions>
      <v-spacer />
      <v-btn
        @click="selectedActivity && $emit('selectedActivity', selectedActivity)"
        color="primary"
        variant="tonal"
        >Run Activity</v-btn
      >
    </v-card-actions>
  </v-card>
</template>

<script setup lang="ts">
import type { Action, Activity, Player } from "@/services/types";
import {
  ACTION_MOTIVATION_MOVE,
  ACTION_POINTS_MOVE,
  costDescription,
  cycleCost,
  getCityActivities,
} from "@/services/cities";
import { computed, ref } from "vue";
import CardSubtitle from "@/components/game/turn/CardSubtitle.vue";

interface Props {
  player: Player;
  action: Action;
}

const props = defineProps<Props>();

defineEmits<{
  (e: "selectedActivity", activity: Activity): void;
}>();

const selectedActivity = ref<Activity>();

function activityLabel(activity: Activity): string {
  return `${activity.name} (${costDescription(activity, props.player)})`;
}

function selectableActivity(a: Activity): boolean {
  if (a.type !== "cycle") {
    return true;
  }

  return canAffordToCycle();
}

function canAffordToCycle(): boolean {
  const { player } = props;
  const { motivation, points } = cycleCost(player);
  return player.motivation >= motivation && player.points >= points;
}

const cityActivities = computed(() => {
  const city = props.player.currentCity;
  return getCityActivities(city!);
});

const actionText = computed(() => {
  const {
    action: { text, type, effect },
  } = props;

  const move =
    type === "motivation" ? ACTION_MOTIVATION_MOVE : ACTION_POINTS_MOVE;
  const smiley = effect === "decrease" ? "ðŸ˜­" : "ðŸ˜Ž";

  let fullText = text;
  fullText += "<br /><br />";
  fullText += `${smiley} Your ${type} ${effect}${
    type === "points" ? "" : "s"
  } by ${move}.`;

  return fullText;
});
</script>
