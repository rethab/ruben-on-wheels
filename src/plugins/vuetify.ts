import "vuetify/styles";
import { createVuetify } from "vuetify";
import { aliases, mdi } from "vuetify/iconsets/mdi";
import * as components from "vuetify/components";
import * as directives from "vuetify/directives";
import "@mdi/font/css/materialdesignicons.css";

export default createVuetify({
  icons: { defaultSet: "mdi", aliases, sets: { mdi } },
  components,
  directives,
});

// Export for test.
export { components, directives };
