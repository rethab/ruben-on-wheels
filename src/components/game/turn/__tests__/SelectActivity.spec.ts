import { describe, it, expect } from "vitest";

import { mount } from "@vue/test-utils";
import SelectActivity from "../SelectActivity.vue";

import vuetify from "@/plugins/vuetify";
import type { Action } from "../../../../services/types";

function mountComponent(action: Action) {
  return mount(SelectActivity as any, {
    props: {
      player: {
        name: "Alice",
        currentCity: "Zurich",
        motivation: 13,
        points: 143,
      },
      action,
    },
    global: {
      plugins: [vuetify],
    },
  });
}

describe("SelectActivity", () => {
  const testAction: Action = {
    text: "sample action text",
    type: "motivation",
    effect: "increase",
  };

  it("renders full component: motivation increase", () => {
    const wrapper = mountComponent(testAction);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders full component: points decrease", () => {
    const wrapper = mountComponent({
      ...testAction,
      type: "points",
      effect: "decrease",
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
