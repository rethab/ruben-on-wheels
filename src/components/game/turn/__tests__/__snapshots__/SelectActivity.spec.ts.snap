// Vitest Snapshot v1

exports[`SelectActivity > renders full component: motivation increase 1`] = `
"<div class=\\"v-card v-theme--light v-card--density-default v-card--variant-elevated\\">
  <!---->
  <div class=\\"v-card__loader\\">
    <div class=\\"v-progress-linear v-theme--light\\" style=\\"height: 0px; --v-progress-linear-height: 2px;\\" role=\\"progressbar\\" aria-hidden=\\"true\\" aria-valuemin=\\"0\\" aria-valuemax=\\"100\\">
      <!---->
      <div class=\\"v-progress-linear__background\\" style=\\"width: 100%;\\"></div>
      <transition-stub name=\\"fade-transition\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
        <div class=\\"v-progress-linear__indeterminate\\">
          <div class=\\"v-progress-linear__indeterminate long\\"></div>
          <div class=\\"v-progress-linear__indeterminate short\\"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class=\\"v-card-title\\">Alice's turn</div>
  <div class=\\"v-card-subtitle\\"><span class=\\"\\"> Points: 143, Motivation: 13, City: Zurich</span></div>
  <div class=\\"v-card-text\\">
    <p>sample action text<br><br>ðŸ˜Ž Your motivation increases by 5.</p>
    <p class=\\"my-5\\">How do you want to proceed?</p>
    <div class=\\"v-input v-input--horizontal v-input--density-default v-radio-group\\">
      <!---->
      <div class=\\"v-input__control\\">
        <!---->
        <div class=\\"v-selection-control-group\\" aria-labelled-by=\\"radio-group-0\\" role=\\"radiogroup\\">
          <div class=\\"v-selection-control v-selection-control--density-default v-radio\\">
            <div class=\\"v-selection-control__wrapper\\">
              <!---->
              <div class=\\"v-selection-control__input\\"><i class=\\"mdi-radiobox-blank mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i><input id=\\"input-3\\" aria-readonly=\\"false\\" type=\\"radio\\" name=\\"radio-group-0\\" value=\\"Visit Laurie &amp; Reto\\">
                <!---->
              </div>
            </div><label class=\\"v-label v-label--clickable\\" for=\\"input-3\\">
              <!---->Visit Laurie &amp; Reto (++ motivation)
            </label>
          </div>
          <div class=\\"v-selection-control v-selection-control--density-default v-radio\\">
            <div class=\\"v-selection-control__wrapper\\">
              <!---->
              <div class=\\"v-selection-control__input\\"><i class=\\"mdi-radiobox-blank mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i><input id=\\"input-4\\" aria-readonly=\\"false\\" type=\\"radio\\" name=\\"radio-group-0\\" value=\\"Cycle up the Uetliberg\\">
                <!---->
              </div>
            </div><label class=\\"v-label v-label--clickable\\" for=\\"input-4\\">
              <!---->Cycle up the Uetliberg (++ points)
            </label>
          </div>
          <div class=\\"v-selection-control v-selection-control--density-default v-radio\\">
            <div class=\\"v-selection-control__wrapper\\">
              <!---->
              <div class=\\"v-selection-control__input\\"><i class=\\"mdi-radiobox-blank mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i><input id=\\"input-5\\" aria-readonly=\\"false\\" type=\\"radio\\" name=\\"radio-group-0\\" value=\\"Take a nap\\">
                <!---->
              </div>
            </div><label class=\\"v-label v-label--clickable\\" for=\\"input-5\\">
              <!---->Take a nap (++ points)
            </label>
          </div>
          <div class=\\"v-selection-control v-selection-control--density-default v-radio\\">
            <div class=\\"v-selection-control__wrapper\\">
              <!---->
              <div class=\\"v-selection-control__input\\"><i class=\\"mdi-radiobox-blank mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i><input id=\\"input-6\\" aria-readonly=\\"false\\" type=\\"radio\\" name=\\"radio-group-0\\" value=\\"Have your bike cleaned\\">
                <!---->
              </div>
            </div><label class=\\"v-label v-label--clickable\\" for=\\"input-6\\">
              <!---->Have your bike cleaned (++ motivation)
            </label>
          </div>
          <div class=\\"v-selection-control v-selection-control--disabled v-selection-control--density-default v-radio\\">
            <div class=\\"v-selection-control__wrapper\\">
              <!---->
              <div class=\\"v-selection-control__input\\"><i class=\\"mdi-radiobox-blank mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i><input disabled=\\"\\" id=\\"input-7\\" aria-readonly=\\"false\\" type=\\"radio\\" name=\\"radio-group-0\\" value=\\"Cycle to the next city\\">
                <!---->
              </div>
            </div><label class=\\"v-label v-label--clickable\\" for=\\"input-7\\">
              <!---->Cycle to the next city (-395 points, -15 motivation)
            </label>
          </div>
        </div>
      </div>
      <!---->
      <div class=\\"v-input__details\\">
        <transition-group-stub name=\\"slide-y-transition\\" tag=\\"div\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\" class=\\"v-messages\\">
          <!---->
        </transition-group-stub>
        <!---->
      </div>
    </div>
  </div>
  <div class=\\"v-card-actions\\">
    <div class=\\"flex-grow-1\\"></div><button type=\\"button\\" class=\\"v-btn v-theme--light text-primary v-btn--density-default v-btn--size-default v-btn--variant-tonal\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
      <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\">Run Activity</span>
      <!---->
      <!---->
    </button>
  </div>
  <!---->
  <!----><span class=\\"v-card__underlay\\"></span>
</div>"
`;

exports[`SelectActivity > renders full component: points decrease 1`] = `
"<div class=\\"v-card v-theme--light v-card--density-default v-card--variant-elevated\\">
  <!---->
  <div class=\\"v-card__loader\\">
    <div class=\\"v-progress-linear v-theme--light\\" style=\\"height: 0px; --v-progress-linear-height: 2px;\\" role=\\"progressbar\\" aria-hidden=\\"true\\" aria-valuemin=\\"0\\" aria-valuemax=\\"100\\">
      <!---->
      <div class=\\"v-progress-linear__background\\" style=\\"width: 100%;\\"></div>
      <transition-stub name=\\"fade-transition\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
        <div class=\\"v-progress-linear__indeterminate\\">
          <div class=\\"v-progress-linear__indeterminate long\\"></div>
          <div class=\\"v-progress-linear__indeterminate short\\"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class=\\"v-card-title\\">Alice's turn</div>
  <div class=\\"v-card-subtitle\\"><span class=\\"\\"> Points: 143, Motivation: 13, City: Zurich</span></div>
  <div class=\\"v-card-text\\">
    <p>sample action text<br><br>ðŸ˜­ Your points decrease by 20.</p>
    <p class=\\"my-5\\">How do you want to proceed?</p>
    <div class=\\"v-input v-input--horizontal v-input--density-default v-radio-group\\">
      <!---->
      <div class=\\"v-input__control\\">
        <!---->
        <div class=\\"v-selection-control-group\\" aria-labelled-by=\\"radio-group-0\\" role=\\"radiogroup\\">
          <div class=\\"v-selection-control v-selection-control--density-default v-radio\\">
            <div class=\\"v-selection-control__wrapper\\">
              <!---->
              <div class=\\"v-selection-control__input\\"><i class=\\"mdi-radiobox-blank mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i><input id=\\"input-3\\" aria-readonly=\\"false\\" type=\\"radio\\" name=\\"radio-group-0\\" value=\\"Visit Laurie &amp; Reto\\">
                <!---->
              </div>
            </div><label class=\\"v-label v-label--clickable\\" for=\\"input-3\\">
              <!---->Visit Laurie &amp; Reto (++ motivation)
            </label>
          </div>
          <div class=\\"v-selection-control v-selection-control--density-default v-radio\\">
            <div class=\\"v-selection-control__wrapper\\">
              <!---->
              <div class=\\"v-selection-control__input\\"><i class=\\"mdi-radiobox-blank mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i><input id=\\"input-4\\" aria-readonly=\\"false\\" type=\\"radio\\" name=\\"radio-group-0\\" value=\\"Cycle up the Uetliberg\\">
                <!---->
              </div>
            </div><label class=\\"v-label v-label--clickable\\" for=\\"input-4\\">
              <!---->Cycle up the Uetliberg (++ points)
            </label>
          </div>
          <div class=\\"v-selection-control v-selection-control--density-default v-radio\\">
            <div class=\\"v-selection-control__wrapper\\">
              <!---->
              <div class=\\"v-selection-control__input\\"><i class=\\"mdi-radiobox-blank mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i><input id=\\"input-5\\" aria-readonly=\\"false\\" type=\\"radio\\" name=\\"radio-group-0\\" value=\\"Take a nap\\">
                <!---->
              </div>
            </div><label class=\\"v-label v-label--clickable\\" for=\\"input-5\\">
              <!---->Take a nap (++ points)
            </label>
          </div>
          <div class=\\"v-selection-control v-selection-control--density-default v-radio\\">
            <div class=\\"v-selection-control__wrapper\\">
              <!---->
              <div class=\\"v-selection-control__input\\"><i class=\\"mdi-radiobox-blank mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i><input id=\\"input-6\\" aria-readonly=\\"false\\" type=\\"radio\\" name=\\"radio-group-0\\" value=\\"Have your bike cleaned\\">
                <!---->
              </div>
            </div><label class=\\"v-label v-label--clickable\\" for=\\"input-6\\">
              <!---->Have your bike cleaned (++ motivation)
            </label>
          </div>
          <div class=\\"v-selection-control v-selection-control--disabled v-selection-control--density-default v-radio\\">
            <div class=\\"v-selection-control__wrapper\\">
              <!---->
              <div class=\\"v-selection-control__input\\"><i class=\\"mdi-radiobox-blank mdi v-icon notranslate v-theme--light v-icon--size-default\\" aria-hidden=\\"true\\"></i><input disabled=\\"\\" id=\\"input-7\\" aria-readonly=\\"false\\" type=\\"radio\\" name=\\"radio-group-0\\" value=\\"Cycle to the next city\\">
                <!---->
              </div>
            </div><label class=\\"v-label v-label--clickable\\" for=\\"input-7\\">
              <!---->Cycle to the next city (-395 points, -15 motivation)
            </label>
          </div>
        </div>
      </div>
      <!---->
      <div class=\\"v-input__details\\">
        <transition-group-stub name=\\"slide-y-transition\\" tag=\\"div\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\" class=\\"v-messages\\">
          <!---->
        </transition-group-stub>
        <!---->
      </div>
    </div>
  </div>
  <div class=\\"v-card-actions\\">
    <div class=\\"flex-grow-1\\"></div><button type=\\"button\\" class=\\"v-btn v-theme--light text-primary v-btn--density-default v-btn--size-default v-btn--variant-tonal\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
      <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\">Run Activity</span>
      <!---->
      <!---->
    </button>
  </div>
  <!---->
  <!----><span class=\\"v-card__underlay\\"></span>
</div>"
`;
